File: py/essential/SubtreeAnotherTree.py
# https://leetcode.com/problems/subtree-of-another-tree/?envType=problem-list-v2&envId=oizxjoit
# Blind

from typing import Optional

from tree_node import *


class Solution:
    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:
        if not s:
            return False
        if self.isSameTree(s, t):
            return True
        return self.isSubtree(s.left, t) or self.isSubtree(s.right, t)

    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        if p and q:
            return (
                p.val == q.val
                and self.isSameTree(p.left, q.left)
                and self.isSameTree(p.right, q.right)
            )
        return p is q  # either is null or both


s = Solution()
node1 = TreeNode(1)
node2 = TreeNode(1)
node1.left = node2
print(s.isSubtree(node1, TreeNode(1)))

File: src/DFS/SubtreeAnotherTree.java
package DFS;

// https://leetcode.com/problems/subtree-of-another-tree/
public class SubtreeAnotherTree {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public boolean isSubtree(TreeNode s, TreeNode t) {
        if (s == null) {
            return false;
        }

        if (s.val == t.val) {
            return isSame(s, t);
        } else {
            return isSubtree(s.left, t) || isSubtree(s.right, t);
        }
    }

    private boolean isSame(TreeNode s, TreeNode t) {
        if (s == null && t == null) return true;
        if (s == null || t == null) return false;

        if (s.val != t.val) {
            return false;
        } else {
            return isSame(s.left, t.left) && isSame(s.right, t.right);
        }
    }

    public void initialize() {
        TreeNode node1 = new TreeNode(4);
        TreeNode node2 = new TreeNode(2);
        TreeNode node3 = new TreeNode(6);
        TreeNode node4 = new TreeNode(1);
        TreeNode node5 = new TreeNode(3);
        TreeNode node6 = new TreeNode(5);

        TreeNode node7 = new TreeNode(7);

        node1.left = node2;
        node1.right = node3;

        node2.left = node4;
        node2.right = node5;

        node3.left = node6;

        //node3.right = node7;

        System.out.println("IS subtree: " + isSubtree(node1, node2));
    }

    public static void main(String[] args) {
        SubtreeAnotherTree r = new SubtreeAnotherTree();
        r.initialize();
    }
}
